src/analysis/refiner.jl

```diff
@@ -179,8 +195,27 @@ function tfunc(::Union{Val{Core.Intrinsics.mul_float}, Val{Core.Intrinsics.mul_i
         return Incidence(builtin_math_tfunc(Core.Intrinsics.mul_float, a.typ, b), a.row * b.val)
     end
     rrow = _zero_row()
-    for i in Iterators.flatten((rowvals(a.row), rowvals(b.row)))
-        rrow[i] = nonlinear
+    ia = rowvals(a.row)
+    ib = rowvals(b.row)
+    time_dependent = in(1, ia) || in(1, ib)
+    states = union(filter(≠(1), ia), filter(≠(1), ib))
+    for i in Iterators.flatten((ia, ib))
+        x = a.row[i]
+        y = b.row[i]
+        if x ≠ 0 && y ≠ 0
+            val = nonlinear
+        else
+            val = x * y
+            if isa(val, Float64)
+                if i == 1 && !isempty(states) # time; state-dependent factors will appear after distribution
+                    val = Linearity(; time_dependent = false, state_dependent = true, nonlinear = false)
+                elseif i > 1 # state
+                    state_dependent = any(≠(i), states)
```

Comment:
This should probably be cross-comparing for better precision, specifically for equations like `(u[1] + u[2])*(::Float64)` and `(u[1] + u[2]) * t`:
```julia
julia> DAECompiler.tfunc(Val(Core.Intrinsics.mul_float), (@incidence u₁ + u₂), DAECompiler.Incidence(Float64)).row
4611686018427387903-element SparseVector{Union{Float64, Linearity}, Int64} with 2 stored entries:
  [2]  =  Linearity(false, true, false)
  [3]  =  Linearity(false, true, false)

julia> DAECompiler.tfunc(Val(Core.Intrinsics.mul_float), (@incidence u₁ + u₂), (@incidence t)).row
4611686018427387903-element SparseVector{Union{Float64, Linearity}, Int64} with 3 stored entries:
  [1]  =  Linearity(false, true, false)
  [2]  =  Linearity(true, true, false)
  [3]  =  Linearity(true, true, false)
```

Reply:
Good catch, thanks.